crates/bevy_pbr/src/material.rs

```diff
@@ -1,321 +1,312 @@
+        let mut descriptor = self.mesh_pipeline.specialize(key.0);
+        if let Some(vertex_shader) = &self.vertex_shader {
+            descriptor.vertex.shader = vertex_shader.clone();
+        }
+
+        if let Some(fragment_shader) = &self.fragment_shader {
+            descriptor.fragment.as_mut().unwrap().shader = fragment_shader.clone();
+        }
+        descriptor.layout = Some(vec![
+            self.mesh_pipeline.view_layout.clone(),
+            self.material_layout.clone(),
+            self.mesh_pipeline.mesh_layout.clone(),
+        ]);
+
+        M::specialize(key.1, &mut descriptor);
+        descriptor
     }
 }
 
-/// The GPU representation of a [`StandardMaterial`].
-#[derive(Debug, Clone)]
-pub struct GpuStandardMaterial {
-    /// A buffer containing the [`StandardMaterialUniformData`] of the material.
-    pub buffer: Buffer,
-    /// The bind group specifying how the [`StandardMaterialUniformData`] and
-    /// all the textures of the material are bound.
-    pub bind_group: BindGroup,
-    pub has_normal_map: bool,
-    pub flags: StandardMaterialFlags,
-    pub base_color_texture: Option<Handle<Image>>,
-    pub alpha_mode: AlphaMode,
+impl<M: SpecializedMaterial> FromWorld for MaterialPipeline<M> {
+    fn from_world(world: &mut World) -> Self {
+        let asset_server = world.get_resource::<AssetServer>().unwrap();
+        let render_device = world.get_resource::<RenderDevice>().unwrap();
+        let material_layout = M::bind_group_layout(render_device);
+
+        MaterialPipeline {
+            mesh_pipeline: world.get_resource::<MeshPipeline>().unwrap().clone(),
+            material_layout,
+            vertex_shader: M::vertex_shader(asset_server),
+            fragment_shader: M::fragment_shader(asset_server),
+            marker: PhantomData,
+        }
+    }
 }
 
-impl RenderAsset for StandardMaterial {
-    type ExtractedAsset = StandardMaterial;
-    type PreparedAsset = GpuStandardMaterial;
-    type Param = (
-        SRes<RenderDevice>,
-        SRes<PbrPipeline>,
-        SRes<RenderAssets<Image>>,
-    );
+type DrawMaterial<M> = (
+    SetItemPipeline,
+    SetMeshViewBindGroup<0>,
+    SetMaterialBindGroup<M, 1>,
+    SetMeshBindGroup<2>,
+    DrawMesh,
+);
 
-    fn extract_asset(&self) -> Self::ExtractedAsset {
-        self.clone()
+pub struct SetMaterialBindGroup<M: SpecializedMaterial, const I: usize>(PhantomData<M>);
+impl<M: SpecializedMaterial, const I: usize> EntityRenderCommand for SetMaterialBindGroup<M, I> {
+    type Param = (SRes<RenderAssets<M>>, SQuery<Read<Handle<M>>>);
+    fn render<'w>(
+        _view: Entity,
+        item: Entity,
+        (materials, query): SystemParamItem<'w, '_, Self::Param>,
+        pass: &mut TrackedRenderPass<'w>,
+    ) -> RenderCommandResult {
+        let material_handle = query.get(item).unwrap();
+        let material = materials.into_inner().get(material_handle).unwrap();
+        pass.set_bind_group(I, M::bind_group(material), &[]);
```

Comment:
it would be nice if the material could specify dynamic offsets so that all material uniforms could live in a single buffer.

Reply:
I think this is desirable / doable, either by exposing a `Material::draw_command(material)` default function or a specific `Material::bind_group_offsets(material) -> usize`. The harder part is ensuring the buffer(s) are properly written to each frame (which requires something extra _after_ the per-material prepare logic runs applying it).

Reply:
Added `Material::dynamic_uniform_indices`, which defaults to an empty list.